cargo-pgrx/src/command/upgrade.rs

```diff
@@ -0,0 +1,216 @@
+        match self {
+            cargo_edit::Source::Registry(reg) => reg.version = new_version.as_ref().to_string(),
+            cargo_edit::Source::Path(path) => path.version = Some(new_version.as_ref().to_string()),
+            cargo_edit::Source::Git(git) => git.version = Some(new_version.as_ref().to_string()),
+            cargo_edit::Source::Workspace(_ws) => {
+                error!(
+                    "Cannot upgrade the version of \
+                    a package because it is set in the \
+                    workspace, please run `cargo pgrx \
+                    upgrade` in the workspace directory."
+                );
+            }
+        }
+    }
+    fn get_version(&self) -> Option<&String> {
+        match self {
+            cargo_edit::Source::Registry(reg) => Some(&reg.version),
+            cargo_edit::Source::Path(path) => path.version.as_ref(),
+            cargo_edit::Source::Git(git) => git.version.as_ref(),
+            cargo_edit::Source::Workspace(_ws) => {
+                error!(
+                    "Cannot fetch the version of \
+                    a package because it is set in the \
+                    workspace, please run `cargo pgrx \
+                    upgrade` in the workspace directory."
+                );
+                None
+            }
+        }
+    }
+}
+fn replace_version<S>(
+    new_version: &str,
+    crate_root: &Path,
+    key: &mut toml_edit::KeyMut,
+    dep: &mut toml_edit::Item,
+    mut parsed_dep: Dependency,
+    mut source: S,
+) -> eyre::Result<()>
+where
+    S: Clone + DependencySource,
+    cargo_edit::Source: From<S>,
+{
+    let dep_name = key.get();
+    let ver_maybe = source.get_version();
+    match ver_maybe {
+        Some(v) => {
+            debug!("{dep_name} version is {v}")
+        }
+        None => return Err(eyre!("No version field for {dep_name}, cannot upgrade.")),
+    }
+    source.set_version(new_version);
+    parsed_dep = parsed_dep.set_source(source);
+
+    parsed_dep.update_toml(crate_root, key, dep);
+    Ok(())
+}
+impl Upgrade {
+    fn update_dep(
+        &self,
+        path: &PathBuf,
+        mut key: KeyMut,
+        dep: &mut toml_edit::Item,
+    ) -> eyre::Result<()> {
+        let dep_name_string = key.get().to_string();
+        let dep_name = dep_name_string.as_str();
+        let parsed_dep: Dependency = match Dependency::from_toml(path.as_path(), dep_name, dep) {
+            Ok(dependency) => dependency,
+            Err(e) => {
+                return Err(eyre!(
+                    "Could not parse dependency \
+                entry for {dep_name} due to error: {e}"
+                ))
+            }
+        };
+        let reg_url = registry_url(path, parsed_dep.registry())
+            .map_err(|e| eyre!("Unable to fetch registry URL for path: {e}"))?;
+        let target_version = match self.target_version {
+            Some(ref ver) => Some(ver.clone()),
```

Comment:
this will happily let the user choose a version that doesn't exist:
```
$ cargo run --manifest-path ../pgrx/Cargo.toml --bin cargo-pgrx pgrx upgrade -t 0.12.0
$ cargo check
error: failed to select a version for the requirement `pgrx = "^0.12.0"`
candidate versions found which didn't match: 0.12.0-alpha.1, 0.12.0-alpha.0, 0.11.4, ...
location searched: crates.io index
required by package `example v0.1.0 (/home/jyn/work/example)`
if you are looking for the prerelease package it needs to be specified explicitly
    pgrx = { version = "0.12.0-alpha.1" }
```
can we add a check that it's in the registry and error out if not? that way we don't write files to disk that won't compile.

Reply:
That sounds reasonable, I'll give it a look.

Reply:
Validating this would require yet another dependency, namely [crates_index](https://docs.rs/crates-index/latest/crates_index/). It might be worth it, though.

Reply:
Seems like overkill to ask the registry for this. I don't feel that strongly about it, but I think it's fine to just do what the user asked. (They can run again if they messed up the first time)

Reply:
So I think we probably _should_ make it smarter, but this might be the wrong direction, so I think we can defer making it smarter until we get some user experience reports.

Reply:
Fair, leaving this alone for now.